src/main/java/fi/helsinki/cs/tmc/core/util/Folders.java

```diff
@@ -10,13 +10,19 @@
     static Path tmp = null;
 
     public static Path tempFolder() {
-        if (tmp == null) {
-            try {
-                tmp = Files.createTempDirectory("tmc-core");
-            } catch (IOException e) {
-                tmp = Paths.get(System.getProperty("java.io.tmpdir"));
+        if (tmp != null) {
+            return tmp;
+        }
+        synchronized (Folders.class) {
+            if (tmp == null) {
```

Comment:
Well this code is called from multiple threads. The first check for tmps non-nullness is not within the mutex. This means that multiple threads can pass over it and then one of the threads takes the mutex. If we don't check nullness again inside mutex it would make next thread taking mutex to execute the null cases code again.    
The non-nullness check outside of mutex is just optimisation: This way after initialisation of tmp, no more locking is needed.